### PR TITLE
feat!: Add collapsible menu to Header

### PR DESCRIPTION
**Pull request by alimpens**
# Describe the pull request

Thank you for contributing to the project!
Please use this template to help us handle your PR smoothly.

## What

It adds a collapsible menu to Header and changes the API accordingly. 

## Why

Header was incomplete, we couldn't make a functional Header yet.

## How

It's quite a rework of the old component. It adds a menu button to the component, and adds children for the content of the collapsible menu.

## Checklist

Before submitting your pull request, please ensure you have done the following. Check each checkmark if you have done so or if it wasn't necessary:

- [x] Add or update unit tests
- [x] Add or update documentation
- [x] Add or update stories
- [x] Add or update exports in index.\* files
- [x] Start the PR title with a Conventional Commit prefix, [as explained here](https://github.com/Amsterdam/design-system/blob/main/documentation/publishing.md?plain=1#L11).

## Additional notes

- Header has a NarrowScreenOnlyGridCell subcomponent, maybe we can improve that name
- Header is supposed to get extra block space. `ams.space.grid.md` was chosen, but that looks awkward imo. I chose `ams.space.grid.sm` now, but that's something we should discuss.
- Should the collapsible menu have fixed top and bottom spacing?
- Spacing in general is something which should be checked by design, it wasn't always clear what it should be.

### CODE REVIEW

File: packages/react/src/Header/Header.tsx [hunk 1]
```diff
@@ -4,61 +4,145 @@
...
+        </div>
+        {(children || menuItems) && (
+          <nav className="ams-header__navigation" aria-labelledby="primary-navigation">
+            <h2 id="primary-navigation" className="ams-visually-hidden">
+              {navigationLabel}
+            </h2>
+
+            {/* The logo and app name section is recreated here, to make sure the page menu breaks at the right spot */}
+            <div className="ams-header__brand-section ams-header__brand-section--hidden" aria-hidden>
+              <div className="ams-header__logo-link">
```
Comment (dlnr) [inline packages/react/src/Header/Header.tsx:86]: Why do you have the logo a second time?

File: proprietary/tokens/src/components/ams/header.tokens.json [hunk 2]
```diff
@@ -1,13 +1,51 @@
...
+        }
+      },
+      "menu": {
+        "column-gap": { "value": "{ams.space.grid.md}" },
+        "row-gap": { "value": "{ams.space.grid.xs}" },
+        "item": {
+          "color": { "value": "{ams.link-appearance.color}" },
+          "column-gap": { "value": "{ams.space.xs}" },
+          "font-family": { "value": "{ams.text.font-family}" },
+          "font-size": { "value": "{ams.text.level.6.font-size}" },
```
Comment (dlnr) [inline proprietary/tokens/src/components/ams/header.tokens.json:28]: I've made a remark on level 6 in [the ticket](https://gemeente-amsterdam.atlassian.net/browse/DES-884), it is weird of the Header font size is significantly smaller than the text right below.

File: proprietary/tokens/src/components/ams/logo.tokens.json [hunk 3]
```diff
@@ -1,8 +1,9 @@
 {
   "ams": {
     "logo": {
-      "block-size": { "value": "2.5rem" },
+      "block-size": { "value": "{ams.space.grid.md}" },
```
Comment (dlnr) [inline proprietary/tokens/src/components/ams/logo.tokens.json:4]: Why use (grid) spacing for block size?

File: packages/react/src/Icon/Icon.tsx [hunk 4]
```diff
@@ -7,15 +7,15 @@
...
+import type { ForwardedRef, HTMLAttributes, ReactNode } from 'react'
 
 export type IconProps = {
   /** The size of the icon. Corresponds with the text levels. */
   size?: 'level-3' | 'level-4' | 'level-5' | 'level-6'
   /** Whether the icon container should be made square. */
   square?: boolean
   /** The component rendering the icon’s markup. */
-  svg: Function
+  svg: Function | ReactNode
```
Comment (dlnr) [inline packages/react/src/Icon/Icon.tsx:18]: Now you can pass any component to Icon, you also could've used HeadMenuIcon.svg or prevented Menu.svg from combining paths.

File: packages/react/src/Header/HeaderMenuLink.tsx
Comment (dlnr) [inline packages/react/src/Header/HeaderMenuLink.tsx:1]: No tests for child components? We usually do add them.

File: storybook/src/components/Avatar/Avatar.stories.tsx
Comment (dlnr) [inline storybook/src/components/Avatar/Avatar.stories.tsx:1]: There is some demand for an avatar in the Header. Why did you remove this?

File: storybook/src/pages/amsterdam/common/AppHeader.tsx [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
+      </Header.MenuLink>,
+      <Header.MenuLink key="2" href="#">
+        Mijn Amsterdam
+      </Header.MenuLink>,
+      <Header.MenuLink key="3" href="#" fixed>
+        Zoeken
+      </Header.MenuLink>,
+    ]}
+  >
+    <LinkList className="ams-mb--lg">
```
Comment (dlnr) [inline storybook/src/pages/amsterdam/common/AppHeader.tsx:18]: Duplicate content might be an issue for public facing websites, especially when it's the main navigation component.

File: packages/css/src/components/header/header.scss [hunk 6]
```diff
@@ -4,83 +4,201 @@
...
 .ams-header__logo-link {
-  flex: none;
   outline-offset: var(--ams-header-logo-link-outline-offset);
 }
 
-.ams-header__links {
-  display: none;
+/* TODO Remove after updating Header line heights in DES-973. */
+.ams-heading.ams-header__app-name {
+  line-height: 1.335;
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/css/src/components/header/header.scss:34]: I'm not sure it's supposed to? The intention here was to match the line height suggested for level 5 [here](https://gemeente-amsterdam.atlassian.net/browse/DES-973) (which is 1.35 btw, I'll change that).

Can you weigh in on what's supposed to happen here @VincentSmedinga?

File: packages/react/src/Header/Header.tsx (hunk 1 above)
Comment (alimpens) [reply packages/react/src/Header/Header.tsx:86]: I used a hack to be able to have all header navigation in 1 `nav`. The `nav` contains a recreation of the logo and app name which is placed behind the real logo and app name. This way, you have the correct semantics and you keep the wrapping behaviour as designed.

File: packages/react/src/Header/HeaderMenuLink.tsx
Comment (alimpens) [reply packages/react/src/Header/HeaderMenuLink.tsx:1]: See the checklist in the PR description, I still have to do the tests.

File: packages/react/src/Icon/Icon.tsx (hunk 4 above)
Comment (alimpens) [reply packages/react/src/Icon/Icon.tsx:18]: I think this change is useful for other uses of the Icon component. Passing a function still works, but if you need to pass a React component with props / a classname, you can.

File: proprietary/tokens/src/components/ams/header.tokens.json [hunk 7]
```diff
@@ -1,13 +1,51 @@
...
+        "button": {
+          "label": {
+            "open": {
+              "font-weight": { "value": "{ams.text.font-weight.bold}" }
+            }
+          }
+        }
+      },
+      "menu": {
+        "column-gap": { "value": "{ams.space.grid.md}" },
```
[the remark this answers is not in the document]
Comment (alimpens) [reply proprietary/tokens/src/components/ams/header.tokens.json:22]: Yeah, makes sense imo to not use grid spacing here. Let's handle this in the ticket you made.

File: proprietary/tokens/src/components/ams/header.tokens.json (hunk 2 above)
Comment (alimpens) [reply proprietary/tokens/src/components/ams/header.tokens.json:28]: I agree, let's discuss this with Bas

File: proprietary/tokens/src/components/ams/logo.tokens.json (hunk 3 above)
Comment (alimpens) [reply proprietary/tokens/src/components/ams/logo.tokens.json:4]: Not sure, I copied this from [1664](https://github.com/Amsterdam/design-system/pull/1664). @VincentSmedinga?

File: storybook/src/components/Avatar/Avatar.stories.tsx
Comment (alimpens) [reply storybook/src/components/Avatar/Avatar.stories.tsx:1]: The removed example doesn't make sense. It's part of the header menu and the main navigation, but it's not an interactive item. If we want to have an Avatar in the Header, we should think about its function and implement it properly.

File: storybook/src/pages/amsterdam/common/AppHeader.tsx (hunk 5 above)
Comment (alimpens) [reply storybook/src/pages/amsterdam/common/AppHeader.tsx:18]: Either the links in the header menu or those in the mega menu have `display: none`, which removes them from the accessibility tree.

File: packages/css/src/components/header/header.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlnr) [reply packages/css/src/components/header/header.scss]: When the app name breaks to a second line it looks misaligned next to the logo, I suggested a while ago that we could match it to the logo. But since then the logo is scaling so that would make it harder.

File: packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.test.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import { render } from '@testing-library/react'
+import { createRef } from 'react'
+import { HeaderNarrowScreenOnlyGridCell } from './HeaderNarrowScreenOnlyGridCell'
+import '@testing-library/jest-dom'
+
+describe('HeaderNarrowScreenOnlyGridCell', () => {
```
Comment (VincentSmedinga) [inline packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.test.tsx:6]: We use spaces in the test names.

File: packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.tsx [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import clsx from 'clsx'
+import { ForwardedRef, forwardRef } from 'react'
+import { Grid, GridCellProps } from '../Grid'
+
+export const HeaderNarrowScreenOnlyGridCell = forwardRef(
```
Comment (VincentSmedinga) [inline packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.tsx:5]: `HeaderGridCellForNarrowWindow`? (or `In` instead of `For`)

(We describe the distinction between a screen and a window in the Grid docs.)

File: packages/react/src/common/useMediaQuery.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { useEffect, useState } from 'react'
+
+// TODO: we should set the breakpoint in JS somewhere and render this and the sass variables from that
```
Comment (VincentSmedinga) [inline packages/react/src/common/useMediaQuery.tsx:3]: Could you create a ticket for this?

File: storybook/src/components/Header/Header.docs.mdx [hunk 11]
```diff
@@ -17,33 +17,18 @@ import { StatusBadge } from "../../docs/components/StatusBadge";
 
 ## Examples
```
Comment (VincentSmedinga) [inline storybook/src/components/Header/Header.docs.mdx:18]: TODO Describe examples.

File: proprietary/tokens/src/components/ams/logo.tokens.json (hunk 3 above)
Comment (VincentSmedinga) [reply proprietary/tokens/src/components/ams/logo.tokens.json:4]: It may be a bit of a stretch, but I wanted to express that the logo is exactly as tall as the vertical whitespace above and below it. Another advantage is that it assumes the same clamping.

File: packages/css/src/components/header/header.scss [hunk 12]
```diff
@@ -4,83 +4,207 @@
...
-  row-gap: 1.5rem;
+  grid-area: 1 / 1;
+}
 
-  @media screen and (min-width: $ams-breakpoint-wide) {
-    column-gap: var(--ams-header-column-gap);
-    flex-wrap: nowrap;
-  }
+.ams-header__brand-section--hidden {
+  opacity: 0%;
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/css/src/components/header/header.scss:24]: Stylelint enforces this

File: packages/css/src/components/header/header.scss [hunk 13]
```diff
@@ -4,83 +4,207 @@
...
+}
+
+@mixin reset-list {
+  margin-block: 0;
+  padding-inline-start: 0;
+}
+
+.ams-header__menu {
+  align-items: center;
+  box-sizing: border-box;
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/css/src/components/header/header.scss:54]: Don't think so, removed it

File: packages/css/src/components/header/header.scss [hunk 14]
```diff
@@ -4,83 +4,207 @@
...
 @import "../../common/breakpoint";
+@import "../../common/text-rendering";
 
 .ams-header {
+  display: grid;
+  padding-block: var(--ams-header-padding-block);
+  padding-inline: var(--ams-grid-padding-inline);
+}
+
+.ams-header__brand-section {
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/css/src/components/header/header.scss:15]: Added them, can you check if they make sense?

File: packages/css/src/components/header/header.scss [hunk 15]
```diff
@@ -4,83 +4,207 @@
...
   }
+}
 
-  @media screen and (min-width: $ams-breakpoint-wide) {
-    order: 3;
+.ams-header__menu-item--fixed {
+  display: revert;
+}
+
+@mixin shared-styles {
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/css/src/components/header/header.scss:78]: I like `header-menu-action`!

File: packages/react/src/Header/HeaderMenuLink.test.tsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
+import { render, screen } from '@testing-library/react'
+import { createRef } from 'react'
+import { HeaderMenuLink } from './HeaderMenuLink'
+import '@testing-library/jest-dom'
+
+describe('Breadcrumb link', () => {
```
[the remark this answers is not in the document]
Comment (alimpens) [reply packages/react/src/Header/HeaderMenuLink.test.tsx:6]: 🦅

File: packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.tsx (hunk 9 above)
Comment (alimpens) [reply packages/react/src/Header/HeaderNarrowScreenOnlyGridCell.tsx:5]: It isn't necessarily 'for' or 'in', it only shows on narrow windows... 

`HeaderGridCellNarrowWindowOnly`?

File: proprietary/tokens/src/components/ams/logo.tokens.json (hunk 3 above)
Comment (alimpens) [reply proprietary/tokens/src/components/ams/logo.tokens.json:4]: It depends on what we do with the Header block spacing for me. I think having the Logo be the same height as the space before and after it looks off. If we change that, maybe we should reconsider this value as well.

File: storybook/src/components/Header/Header.docs.mdx (hunk 11 above)
Comment (alimpens) [reply storybook/src/components/Header/Header.docs.mdx:18]: I've added these

File: packages/css/src/components/header/README.md [hunk 17]
```diff
@@ -12,12 +12,14 @@ Includes the name of the application if it is not the general website.
 - The Header is important because it conveys our corporate identity and is the first thing people see.
   Using it consistently helps users recognize and trust the website.
 - It is the same on every page of the application.
```
Comment (alimpens) [inline packages/css/src/components/header/README.md:14]: Also for full page map type of things? The current and new Meldingen design doesn't follow this guideline atm

File: proprietary/tokens/src/components/ams/logo.tokens.json (hunk 3 above)
Comment (VincentSmedinga) [reply proprietary/tokens/src/components/ams/logo.tokens.json:4]: We’ve changed this to use regular spacing.

File: proprietary/tokens/src/components/ams/header.tokens.json (hunk 2 above)
Comment (VincentSmedinga) [reply proprietary/tokens/src/components/ams/header.tokens.json]: We’ve changed this to use Text Level 5.

File: packages/css/src/components/header/README.md (hunk 17 above)
Comment (VincentSmedinga) [reply packages/css/src/components/header/README.md]: We’ve updated the text.

**Review by VincentSmedinga — APPROVED** [review]
Alea iacta est!
